### PR TITLE
Fault-handling (including watchdog)

### PR DESCRIPTION
**Pull request by henryhallam**
This pull req will ultimately address #326 #327 #328 #329 #330.

### CODE REVIEW

**Comment by henryhallam** [comment]
Screaming death: Closes #326 
_exit handler: Closes #327

**Comment by henryhallam** [comment]
Fault handler: Closes #328 
Would be nice to add more trace functionality to this, but it should already tell us much more than "hurr durr system crashed"

@fnoble @gsmcmullin review plz

**Comment by henryhallam** [comment]
@fnoble I'm not going to add fallback_write for all the UARTs right now, but I've put a TODO for that as well as the more extensive stack/state dump.  The only hazard I can think of with this is that the integer-divide-by-zero trapping is new, and perhaps there are some divide-by-zeros occurring occasionally that we didn't notice before, which will now result in a crash.  I think we'd want to know about them, though.  FWIW in single-ended mode with a roof antenna, that is not occurring.

So, OK to merge?

**Comment by fnoble** [comment]
Looks good, just going to run it on HITL for a while

File: src/peripherals/usart.c [hunk 1]
```diff
@@ -184,7 +184,18 @@ void usarts_enable(u32 ftdi_baud, u32 uarta_baud, u32 uartb_baud, bool do_precon
 
   if (do_preconfigure_hooks) {
 
+    /* TODO: Should this really be here? */
     log_info("Piksi Starting...\n");
+    if ((RCC_CSR & 0xFF000000) != RCC_CSR_PINRSTF) {
+      log_info("Reset reason: %02X\n", (unsigned int)(RCC_CSR >> 24));
+      if (RCC_CSR & RCC_CSR_IWDGRSTF)
+        log_warn("Watchdog reset detected.\n");
```
Comment (fnoble) [inline src/peripherals/usart.c]: Make this a `log_error` as discussed

File: src/peripherals/usart.c [hunk 2]
```diff
@@ -184,7 +184,18 @@ void usarts_enable(u32 ftdi_baud, u32 uarta_baud, u32 uartb_baud, bool do_precon
...
+      if (RCC_CSR & RCC_CSR_LPWRRSTF)
+        log_warn("Low power reset detected.\n");
```
Comment (fnoble) [inline src/peripherals/usart.c]: I'd think about making this an error too?

**Comment by fnoble** [comment]
@henryhallam : Could we try running this on HITL with the timeout set much longer (e.g. 20s)? Then we can catch the irrecoverable crashes for now and get this in master before the locking issues are fixed (which might take some time).

Also would appreciate if you could rebase and make those last changes to the `log_warn`s. Then I think this is good to go!

**Comment by henryhallam** [comment]
Done.  WDT hardware period can't be > 32768 ms so I made it 30 seconds and the thread period 15 seconds.  Do you want to run it on the mega HITL or just the regular automatic PR HITL?

**Comment by fnoble** [comment]
@henryhallam : Awesome, I'll set it going on mega HITL now

**Comment by fnoble** [comment]
HITL looks good

**Pull request merged**